packages/logging/src/analytics-helpers.ts

```
@@ -119,4 +159,206 @@ export class ToggleableAnalytics implements MongoshAnalytics {
       }
     }
   }
+
+  flush(callback: (err?: Error | undefined) => void): void {
+    return this._target.flush(callback);
+  }
+}
+
+type ThrottledAnalyticsOptions = {
+  target: MongoshAnalytics;
+  /**
+   * Throttling options. If not provided, throttling is disabled (default: null)
+   */
+  throttle: {
+    /** Allowed events per timeframe number */
+    rate: number;
+    /** Timeframe for throttling in milliseconds (default: 60_000ms) */
+    timeframe?: number;
+    /** Path to persist rpm value to be able to track them between sessions */
+    metadataPath: string;
+    /** Duration in milliseconds in which the lock is considered stale (default: 43_200_000) */
+    lockfileStaleDuration?: number;
+  } | null;
+};
+
+async function lockfile(
+  filepath: string,
+  staleDuration = 43_200_000
+): Promise<() => Promise<void>> {
+  let intervalId: ReturnType<typeof setInterval>;
+  const lockfilePath = `${filepath}.lock`;
+  const unlock = async() => {
+    clearInterval(intervalId);
+    try {
+      return await fs.promises.rmdir(lockfilePath);
+    } catch {
+      // ignore update errors
+    }
+  };
+  try {
+    await fs.promises.mkdir(lockfilePath);
+    // Set up an interval update for lockfile mtime so that if the lockfile is
+    // created by long running process (longer than staleDuration) we make sure
+    // that another process doesn't consider lockfile stale
+    intervalId = setInterval(() => {
+      const now = Date.now();
+      fs.promises.utimes(lockfilePath, now, now).catch(() => {});
+    }, staleDuration / 2);
+    // eslint-disable-next-line chai-friendly/no-unused-expressions
+    intervalId.unref?.();
+    return unlock;
+  } catch (e) {
+    if ((e as any).code !== 'EEXIST') {
+      throw e;
```

does this error have any impact on the user?

---

EEXIST is the only expected error here (at this point most of mongosh startup code made sure that is is accessible), so we probably shouldn't ignore them if they happen

---

Lol, I totally answered incorrectly here 🙈 When we use this method below we actually ignore the errors with a comment: https://github.com/mongodb-js/mongosh/pull/1475/files#diff-fd5487a5c6b6e79731075e82f685921ea435c73c9c7e2869b87c2049de12da61R296-R305